### PR TITLE
[Merged by Bors] - feat(NumberTheory/SmoothNumbers): multiplicativity

### PR DESCRIPTION
**Pull request by rwst**
Prove that the product of two `n`-smooth numbers is an `n`-smooth number. Uses two helper lemmas.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:276]:
```suggestion
lemma primeFactors_subset_of_mem_smoothNumbers {m n : ℕ} (hms : m ∈ n.smoothNumbers) :
    m.primeFactors ⊆ n.primesBelow :=
  have hxle {x : ℕ} (hpf : x ∈ m.primeFactors) (hms : m ∈ n.smoothNumbers) :
      x < n :=
    match mem_smoothNumbers.mp hms with
    | ⟨_, h⟩  => h x (mem_primeFactors_iff_mem_factors.mp hpf)
  fun _ h => mem_primesBelow.mpr ⟨hxle h hms, prime_of_mem_primeFactors h⟩
```
Notes:
- Use implicit variables when it makes sense (by using `{` and `}`  instead of `(` and `)`).
- Why use `h'` or `h2` for hypothesis names when `h` is available?

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:282]:
```suggestion
  ⟨hm, fun _ h => lt_of_mem_primesBelow <| hp <| mem_primeFactors_iff_mem_factors.mpr h ⟩
```

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:280]:
```suggestion
lemma mem_smoothNumbers_of_primeFactors_subset {m n : ℕ} (hm : m ≠ 0)
```

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:352]:
```suggestion
theorem mul_mem_smoothNumbers {m₁ m₂ n : ℕ}
    (hm1 : m₁ ∈ n.smoothNumbers) (hm2 : m₂ ∈ n.smoothNumbers) :
      (m₁ * m₂) ∈ n.smoothNumbers :=
  have hm1' := primeFactors_subset_of_mem_smoothNumbers hm1
  have hm2' := primeFactors_subset_of_mem_smoothNumbers hm2
  mem_smoothNumbers_of_primeFactors_subset (mul_ne_zero hm1.1 hm2.1)
    <| primeFactors_mul hm1.1 hm2.1 ▸ Finset.union_subset hm1' hm2'
```

**Comment by rwst** [comment]
Thanks for the improvements. I'm learning.

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:268]:
```suggestion
/-- The prime factors of an `n'-smooth number are contained in the set of primes below `n`
```

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:278]:
```suggestion
/-- `m` is a `n`-smooth number if all of its prime factors are contained in the set of
```

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:343]:
```suggestion
/-- The product of two `n`-smooth numbers is an `n`-smooth number -/
```

**Review by Rida-Hamadani — COMMENTED** [review]
Thanks! Can you also provide `mem_smoothNumbers_iff_primeFactors_subset`?

**Comment by rwst** [comment]
Here it is!

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:288]:
```suggestion
lemma mem_smoothNumbers_iff_primeFactors_subset {m n : ℕ} :
    m ∈ n.smoothNumbers ↔ m ≠ 0 ∧ m.primeFactors ⊆ n.primesBelow :=
  ⟨fun h => ⟨h.1, primeFactors_subset_of_mem_smoothNumbers h⟩,
  fun h => mem_smoothNumbers_of_primeFactors_subset h.1 h.2⟩
```

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (Rida-Hamadani) [inline Mathlib/NumberTheory/SmoothNumbers.lean:269]:
```suggestion
  if `m` is an `n`-smooth number -/
```

**Review by Rida-Hamadani — APPROVED** [review]

**Comment by Rida-Hamadani** [comment]
LGTM

**Review by grunweg — COMMENTED** [review]
In fact, I took a look and left some stylistic comments.

I am not a number theorist, hence cannot comment on the substance of this PR.

**Comment by grunweg** [comment]
awaiting-author

**Comment by rwst** [comment]
I would like to deal with the comments in a sequential way, not the least in order to preserve the optimization steps that were done.

File: Mathlib/NumberTheory/SmoothNumbers.lean
Comment (rwst) [reply Mathlib/NumberTheory/SmoothNumbers.lean:273]: I agree. Here tactics mode looks better than the clunky match.

**Comment by rwst** [comment]
@MichaelStollBayreuth 
I have flagged the PR as WIP, to signal that there is an alternative (and mathematically cleaner) way to do this. I will now open a new branch for that work. Once that is done and merged, this one should be closed.

**Review by loefflerd — COMMENTED** [review]
LGTM, thanks Ralf and Michael!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
